stable/kommander/templates/kubecost/kubecost-query-stores-configmap.yaml

```
@@ -4,19 +4,48 @@
 {{- if index .Values "kubecost" "cost-analyzer" "thanos" "query" "enabled" }}
 {{- range index .Values "kubecost" "cost-analyzer" "thanos" "query" "serviceDiscoveryFileConfigMaps" }}
 ---
-apiVersion: v1
-kind: ConfigMap
+apiVersion: batch/v1
+kind: Job
 metadata:
-  name: {{ . }}
+  name: create-{{ . }}-configmap
   namespace: {{ $ns }}
   labels:
 {{ $labels }}
   annotations:
-    helm.sh/hook: pre-install
-    helm.sh/hook-delete-policy: before-hook-creation
-data:
-  stores.yaml: |-
-    - targets: []
+    helm.sh/hook: pre-install,pre-upgrade
+    helm.sh/hook-weight: "4"
+    helm.sh/hook-delete-policy: before-hook-creation,hook-succeeded
+spec:
+  template:
+    metadata:
+      name: create-{{ . }}-configmap
+    spec:
+      serviceAccountName: {{ $sa }}
+      restartPolicy: OnFailure
+      containers:
+        - name: kubectl
+          image: bitnami/kubectl:1.16.2
+          command:
+          - sh
+          - "-c"
+          - |
+            /bin/bash <<'EOF'
+            set -o nounset
+            set -o errexit
+            set -o pipefail
+
+            echo "checking if {{ . }} configmap exists"
+
+            RES=$(kubectl get configmap {{ . }} || true)
```

Do you want the script to fail if `kubectl get configmap` fails? If you do, I believe you will need to enable `pipefail` within the subshell. If you just want to ignore a 404 (which is normal, if the configmap wasn't created in the previous version of the Chart, and is to be created on upgrade), then I recommend using `--ignore-not-found=true`.

```suggestion
            RES=$(set -o errexit; set -o pipefail; kubectl get --ignore-not-found=true configmap {{ . }})
```

---

good call! I'll pass in the `ignore-not-found` flag which is cleaner :) (`set -o errexit` and `set -o pipefail` are set at the top of script)

---

I think we were both a little wrong about the bash options :joy:

Bash subshells do not inherit options intuitively.

For years, I had thought that a command-substitution subshell does not inherit `pipefail`, but in fact it does, so it is not necessary to set it in any subshell.

On the other hand, a command-substitution subshell does not inherit `errexit`, so it is necessary to set it in every subshell. (There are two workarounds: Use bash in its non-default POSIX mode, or use bash >= 4.4 and set the `inherit_errexit` option in the containing shell.)

For additional details and some good examples, please see https://dougrichardson.us/2018/08/03/fail-fast-bash-scripting.html

---

I think that, if we don't set `errexit` in the subshell, then, if kubectl throws an error for some reason (e.g. network issue), then the script will drop down to "configmap already exists."